### DISCUSSION TITLE
Smart Meter Interface - Understanding of documentation

### DISCUSSION THREAD

**Discussion by Pewibe** [arendst/Tasmota#12497] [Q&A] [answered]
I like to implement the processing of Smart Meter "EasyMeter Q3A" and edited / compiled Tasmota according to the 1st page of the doc. Next in the doc the chapters 'Descriptor Syntax' and 'Meter Definition' I can understand in general, but it is not clear for me HOW and WHERE to apply all this definitions. 
Is it possible to explain in some sentences which steps have to be done to get the data collection of a Smart Meter ready? If this could be added to the doc it would be perfect.
And a second good idea would be to seperate the doc into 2 parts:
1. What to do if using a predefined Smart Meter (like my EasyMeter Q3A).
2. What to do if an undefined Smart Meter have to be used.
--> Hint: There are some mistakes contained in the Table Of Content of the doc which should be corrected.

**Comment by barbudor** [comment, marked as answer]
> I ... edited / compiled Tasmota according to the 1st page of the doc. 

It is generally a good thing to elaborate on what you have exactly done as"according to the doc" is very vague.

> it is not clear for me HOW and WHERE to apply all this definitions.

Smart Meting relies in the Tasmota Scripting Language.
Did you read the [Scripting-Language](https://tasmota.github.io/docs/Scripting-Language/) of the doc ?

> Is it possible to explain in some sentences which steps have to be done to get the data collection of a Smart Meter ready? If this could be added to the doc it would be perfect.

Tasmota's doc like Tasmota code is OpenSource
You can edit and suggest modifications to the doc by clicking on the pencil icon on the top right

> And a second good idea would be to seperate the doc into 2 parts:
> 
> 1. What to do if using a predefined Smart Meter (like my EasyMeter Q3A).

There are already a few meters that are described..
Feel free to do as suggested  above and add your meter.

> 2. What to do if an undefined Smart Meter have to be used.
>    --> Hint: There are some mistakes contained in the Table Of Content of the doc which should be corrected.

So this is a guessing game ?
You tell us that there are errors but we have to find them by yourself ?
Either you fix them by yourself or you tell us exactly what is wrong.

**Reply by Pewibe** [reply]
@barbudor : To say it generally, the problem is my missing intelligence together with a non-optimal documentation!

> ... edited / compiled Tasmota according to the 1st page of the doc. 
> > It is generally a good thing to elaborate on what you have exactly done as"according to the doc" is very vague
> > > I wrote "according to **the 1st page** of the doc" which describes the changes in 'user_config_override.h' and some #define **in a not-named module**. I named it '1st page' because it has no headline and therefore I cannot name it. After compiling linking I have done nothing else because it is not clear for me to select the predefined EasyMeter Q3A. --> Can somebody tell me how?

> it is not clear for me HOW and WHERE to apply all this definitions.
> > Smart Meting relies in the Tasmota Scripting Language.
Did you read the Scripting-Language of the doc ?
> > > No, I didn't because it was not clear to me that scripting of Smart Meters is the general Tasmota scripting and not a special one. Now it is clear and I shall read it. Thanks!

> Is it possible to explain in some sentences which steps have to be done to get the data collection of a Smart Meter ready? If this could be added to the doc it would be perfect.
> > Tasmota's doc like Tasmota code is OpenSource
You can edit and suggest modifications to the doc by clicking on the pencil icon on the top right
> > > I know that but there are 2 issues: 
a) To edit text you have to understand the whole process and better have some practice first. After I have done what I want, I might be ready for correcting documentation.  
b) I did documentation changes in the past. Not all of them were taken over into official doc. If you feel that your input might be for the waste basket, it makes the author tired. 
--> I believe that first there has to be a 'doc connection' between original author and corrector for co-working.

> And a second good idea would be to seperate the doc into 2 parts:
What to do if using a predefined Smart Meter (like my EasyMeter Q3A).
> > There are already a few meters that are described..
Feel free to do as suggested above and add your meter.
> > > You misunderstood me. My Smart Meter EasyMeter Q3A IS ALREADY defined and part of documentation. But is is not visible for me HOW TO USE this predefined one. And again: Do edit doc for this Smart Meter I have to operate it, first. 

> --> Hint: There are some mistakes contained in the Table Of Content of the doc which should be corrected.
> > So this is a guessing game ?
You tell us that there are errors but we have to find them by yourself ?
Either you fix them by yourself or you tell us exactly what is wrong.
> > > Sorry, it should not target to a guessing game. I wrote 'Table Of Content' and thought that somebody could have a look to the TOC. If I'm the only one who see TOC corrupted, please refer to the following image:
![image](https://user-images.githubusercontent.com/42893315/123632161-e901f880-d817-11eb-8e77-d9c10117f407.png)

**Btw, who is the original author / last large editor of the Smart Meter doc? Is it code producer Gerhard Mutz?**

**Reply by barbudor** [reply]
Sorry if my response offended you but you cannot jump into a project like Tasmota without a minimum of going through the documentation. Tasmota is a community / voluntary supported software which still requires skills to be used. Not a consumer product.

The 1st lines of the Smart Meter page clearly makes reference to Scripting language by usage of `#define USE_SCRIPT` directive.

Apparently, now that you precisely pointed out I can see that the whole page has been damaged by an inappropriate change and need to be cleaned up.
I will look into that from a text formatting point of view only as I am not a SML user myself.
May be the doc will be come more understandable once I have cleanned it up

But generally calling for help without showing what you have already done, doesn't catch much attention. A better baseline is generally elaborating on
- What you have understood
- What you have exactly done (commands, scripts, compile options, ...)
- What you were expecting to get
- What you are effectively getting

It doesn't matter who wrote the original documentation. Many have probably participated and provided changes. If you are such curious to know, don't ask : the answer is in GitHub which tracks all changes.

**Reply by barbudor** [reply]
Doc as been fixed and may be more understandable now

It seems that the Q3A has been added in the docs by @MichiM2 who may be able to help you.
@nagyrobi seems also to have participated on this section.

**Reply by nagyrobi** [reply]
I've spent many hours fixing it at that time. But every once in a while it breaks when a new example is added and markdown syntax is not respected.

**Reply by barbudor** [reply]
@nagyrobi I don't know if you know Q3A but may be you can help @Pewibe ?

**Reply by nagyrobi** [reply]
It was not me who added the Q3A originally.

**Reply by Pewibe** [reply]
Thanks for your discussions and mostly for correcting the doc so quickly. I thinks it's not necessary to support me concerning the EasyMeter Q3A because my problems were generally. 
My erroneous understanding of the documented Smart Meters was that they are a kind of 'implemented and ready to use' devices. So I was looking for a mechanism of 'switch on' the needed Q3A. Now I got it that this Meters are tested and documented scripts. Please apologize my misinterpretations of the Smart Meter Interface.
